PR - feature/BACK-SETUP_CU-28et7jm

antoniokunstek opened the pull request:
Please review current project setup.

antoniokunstek committed:
created env files

antoniokunstek committed:
project settings file

antoniokunstek committed:
created git ignore file

antoniokunstek committed:
installed project dependencies

antoniokunstek committed:
added env files to git ignore

antoniokunstek committed:
Updated git ignore file

antoniokunstek committed:
Initial express application

antoniokunstek committed:
established MongoDB connection

antoniokunstek committed:
created example routes

antoniokunstek committed:
created error classes

antoniokunstek committed:
created mock file with users

antoniokunstek committed:
created MongoDB user model

antoniokunstek committed:
created middleware files

antoniokunstek committed:
deleted unused imports

antoniokunstek commented:
Task linked: [CU-28et7jm Stvaranje novog back-end repozitorija](https://app.clickup.com/t/28et7jm)

mislavmatijevic commented on models/User.js:
```diff
@@ -0,0 +1,77 @@
+		maxlength: 100,
+	},
+	email: {
+		type: String,
```
It should be checked with a Regex. See: https://stackoverflow.com/questions/18022365/mongoose-validate-email-syntax

antoniokunstek replied:
Here I used validator package. See: https://www.npmjs.com/package/validator

mislavmatijevic commented on models/User.js:
```diff
@@ -0,0 +1,77 @@
+		type: String,
+		unique: true,
+		required: [true, 'Please provide email'],
+		minlength: 8,
```
Why would this ever have `minlength`? Pure SHA256 is already 64 chars, let alone salt.

antoniokunstek replied:
Removed :)

mislavmatijevic commented on models/User.js:
```diff
@@ -0,0 +1,77 @@
+
+UserSchema.pre('save', async function () {
+	if (!this.isModified('password')) return;
+	const salt = await bcrypt.genSalt(10);
```
This function would need to be checked. If you generate 10 byte salt, I think hash is 50-ish chars long. It'd be great if you'd check for a better way to use bcrypt and to generate full 71 char passwords.

antoniokunstek replied:
Fixed :)

mislavmatijevic commented on mock/users.json:
```diff
@@ -0,0 +1,45 @@
+    "active": false,
+    "suspended": false,
+    "regDate": "01-01-2000",
+    "lastVisit": "01-01-2000"
```
Maybe the dates could be something more realistic?

antoniokunstek replied:
Fixed :)

mislavmatijevic commented on middleware/authentication.js:
```diff
@@ -0,0 +1,33 @@
+module.exports = {
+	authenticateUser,
+	authorizePermissions,
+};
```
Pretty sure this should also be part of the comment at this stage.

mislavmatijevic replied:
I'm refering to the comment above that takes most of the file, including these two constants.

mislavmatijevic commented on controllers/authController.js:
```diff
@@ -0,0 +1,58 @@
+};
+
+const test = async (req, res) => {
+	res.status(200).json({ msg: 'Route is Up!' });
```
This is just a set-up of the repo, but these return objects ain't gonna cut it. You decide whether to leave them for now. I'd change them to something more appropriate (see: docs).

antoniokunstek replied:
Fixed :)

mislavmatijevic commented on controllers/authController.js:
```diff
@@ -0,0 +1,58 @@
+	//const tokenUser = createTokenUser(user);
+	//attachCookiesToResponse({ res, user: tokenUser });
+
+	res.status(200).json({ user: 'blah-blah' });
```
This ain't right. But you decide whether to leave it for a future ticket or to actually add some sense to this. Or to just wrap everything here in a comment and to call it a day.

mislavmatijevic commented on utils/jwt.js:
```diff
@@ -0,0 +1,29 @@
+const jwt = require('jsonwebtoken');
+
+const createJWT = ({ payload }) => {
+	const token = jwt.sign(payload, process.env.JWT_SECRET, {
```
The variable is called `JWT_KEY` in both '.env' files. Devil is in the details.

mislavmatijevic replied:
The variable in both '.env' files is called `JWT_KEY`, not `JWT_SECRET`. Devil is in the details.

mislavmatijevic commented on utils/jwt.js:
```diff
@@ -0,0 +1,29 @@
+
+const createJWT = ({ payload }) => {
+	const token = jwt.sign(payload, process.env.JWT_SECRET, {
+		expiresIn: process.env.JWT_LIFETIME,
```
Either add this to both .env files or figure something else out.

mislavmatijevic commented on app.js:
```diff
@@ -0,0 +1,56 @@
+
+// middleware
+
+app.set('trust proxy', 1);
```
I'm not really sure we need this? Why do you think we'll be using proxies?

antoniokunstek replied:
I remove it. :)

mislavmatijevic commented on app.js:
```diff
@@ -0,0 +1,56 @@
+app.set('trust proxy', 1);
+app.use(
+	rateLimiter({
+		windowMs: 15 * 60 * 1000,
```
Not really sure what this is?

antoniokunstek replied:
This was part of express-rate-limiter, it is used to limit requests to public APIs.

mislavmatijevic commented on app.js:
```diff
@@ -0,0 +1,56 @@
+
+app.use('/api/v1/auth', authRouter);
+
+app.get('/', (req, res) => {});
```
This is not needed if we will be using routers.

antoniokunstek replied:
Fixed :)

mislavmatijevic commented on app.js:
```diff
@@ -0,0 +1,56 @@
+app.use('/api/v1/auth', authRouter);
+
+app.get('/', (req, res) => {});
+
```
We could use an app.use with an error handling router. But that can be left for another ticket. Just worth noting here.

antoniokunstek replied:
I totally agree, I added it in this ticket. :)

mislavmatijevic commented on config.js:
```diff
@@ -0,0 +1,14 @@
+	HOST: process.env.HOST || 'localhost',
+	PORT: process.env.PORT || 3000,
+	DB: process.env.MONGO_DB || null,
+	JWT_SECRET_KEY: process.env.JWT_KEY || null,
```
Maybe we could export an actual key here instead of a `null`, but, of course, just for testing purposes?

antoniokunstek replied:
I added default key and default json web token lifetime. :)

mislavmatijevic commented on package.json:
```diff
@@ -0,0 +1,48 @@
+    "jsonwebtoken": "^8.5.1",
+    "mongoose": "^6.3.1",
+    "morgan": "^1.10.0",
+    "nodemon": "^2.0.15",
```
This should be a dev dependency!

mislavmatijevic replied:
No idea why fricking GitHub mentions 4 lines. I specifically selected line 44: nodemon.

antoniokunstek replied:
It was a problem with depracated npm install. I moved packages to dev dependencies. :)

mislavmatijevic commented on README.md:
```diff
@@ -1,2 +1,7 @@
 # gym-master-back
```
This could be "# Gym Master - Back-end Repozitorij".
If people are going to read this, they should get a little bit of extra info from this. Maybe something more in the lines of our front-end readme: https://github.com/Toujane/gym-master-front/blob/main/README.md

mislavmatijevic reviewed (changes requested):
A ""few"" comments. Couldn't help myself but to be annoying ¯\_(ツ)_/¯
Only like two or three need actual fixing for future sake.
Btw, I'd prefer commit names to be in capital letters. I like everything standardized so I wrote a doc on that topic.

mislavmatijevic commented on utils/jwt.js:
```diff
@@ -0,0 +1,29 @@
+
+const createJWT = ({ payload }) => {
+	const token = jwt.sign(payload, process.env.JWT_SECRET, {
+		expiresIn: process.env.JWT_LIFETIME,
```
JWT_LIFETIME is undefined. Either add this to both .env files or figure something else out.

antoniokunstek replied:
Fixed :)

mislavmatijevic commented on utils/jwt.js:
```diff
@@ -0,0 +1,29 @@
+const jwt = require('jsonwebtoken');
+
+const createJWT = ({ payload }) => {
+	const token = jwt.sign(payload, process.env.JWT_SECRET, {
```
The variable in both '.env' files is called `JWT_KEY`, not `JWT_SECRET`. Devil is in the details.

antoniokunstek replied:
Fixed :)

antoniokunstek committed:
Updated JWT Names

Changed JWT secret key  and JWT lifetime variable name.

antoniokunstek committed:
Changed API Responses

Changed the way of returning API response. It is changed to a specific way, mentioned in docs.

antoniokunstek committed:
Created Error Handling

Added dependency for async error handling.

antoniokunstek committed:
Deleted Unnecessary Files

Some files were useless.

antoniokunstek committed:
Updated User Model Schema

Changed password min length. Added email validation system.

antoniokunstek committed:
Created Routing System

Added routes for user authentication.

antoniokunstek committed:
Added JWT Token

Added token creation on successful registration and login.

antoniokunstek committed:
Updated Dev Dependencies

Changed some regular dependencies to dev dependencies.

antoniokunstek committed:
Removed Proxy Settings

Removed proxy settings.

mislavmatijevic reviewed (approved):
Great work. Let's get down to business!
